File: pi-lucidia/octavia-operator/packages/blackroad-os-metrics-standalone/stripe/stripe_webhook.py
#!/usr/bin/env python3
"""
Stripe Webhook Handler for BlackRoad OS
Handles payment events and updates sponsor database

Copyright: BlackRoad OS, Inc.
"""

import os
import json
from flask import Flask, request, jsonify
import stripe

app = Flask(__name__)

# Configure Stripe
stripe.api_key = os.getenv('STRIPE_SECRET_KEY')
webhook_secret = os.getenv('STRIPE_WEBHOOK_SECRET')

@app.route('/webhook', methods=['POST'])
def webhook():
    payload = request.data
    sig_header = request.headers.get('Stripe-Signature')

    try:
        event = stripe.Webhook.construct_event(
            payload, sig_header, webhook_secret
        )
    except ValueError as e:
        return jsonify({'error': 'Invalid payload'}), 400
    except stripe.error.SignatureVerificationError as e:
        return jsonify({'error': 'Invalid signature'}), 400

    # Handle the event
    if event['type'] == 'checkout.session.completed':
        session = event['data']['object']
        handle_successful_payment(session)

    elif event['type'] == 'customer.subscription.created':
        subscription = event['data']['object']
        handle_new_subscription(subscription)

    elif event['type'] == 'customer.subscription.deleted':
        subscription = event['data']['object']
        handle_cancelled_subscription(subscription)

    return jsonify({'status': 'success'}), 200

def handle_successful_payment(session):
    # Update sponsor database
    # Send thank you email
    # Add to SPONSORS.md
    pass

def handle_new_subscription(subscription):
    # Add to recurring sponsors list
    # Send welcome email
    pass

def handle_cancelled_subscription(subscription):
    # Update sponsor status
    # Send follow-up email
    pass

if __name__ == '__main__':
    app.run(port=4242)

File: pi-lucidia/octavia-operator/packages/blackroad-stripe-integration/stripe-webhook.ts
import { NextRequest, NextResponse } from 'next/server'
import Stripe from 'stripe'

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY!, {
  apiVersion: '2024-12-18.acacia'
})

const webhookSecret = process.env.STRIPE_WEBHOOK_SECRET!

export async function POST(req: NextRequest) {
  const body = await req.text()
  const signature = req.headers.get('stripe-signature')!

  let event: Stripe.Event

  try {
    event = stripe.webhooks.constructEvent(body, signature, webhookSecret)
  } catch (err) {
    return NextResponse.json(
      { error: 'Webhook signature verification failed' },
      { status: 400 }
    )
  }

  // Handle events
  switch (event.type) {
    case 'checkout.session.completed':
      const session = event.data.object as Stripe.Checkout.Session
      console.log('💰 Payment successful:', session.id)
      // TODO: Provision user access, send email, update database
      break

    case 'customer.subscription.created':
      console.log('🎉 New subscription:', event.data.object.id)
      break

    case 'customer.subscription.updated':
      console.log('📝 Subscription updated:', event.data.object.id)
      break

    case 'customer.subscription.deleted':
      console.log('❌ Subscription cancelled:', event.data.object.id)
      // TODO: Revoke user access
      break

    case 'invoice.payment_succeeded':
      console.log('✅ Invoice paid:', event.data.object.id)
      break

    case 'invoice.payment_failed':
      console.log('⚠️ Payment failed:', event.data.object.id)
      // TODO: Send notification to user
      break
  }

  return NextResponse.json({ received: true })
}
